### PR TITLE
Use sensible maths for `from_rational`

### PR DESCRIPTION
**Pull request by gavofyork**
Old maths had all sorts of small edge cases with rounding which broke expectations. Now we just use the stuff which is already proven.

Note one test in staking was relying on the old buggy maths. It had to be corrected with some dirty magic numbers. Not sure if @kianenigma wants to take a look. A test was added in `per_things` (`from_rational_with_rounding_breakage_2`) to demonstrate that the old numbers for which this tests behaviour relied on are in fact wrong and do definitely break expectations.

Possibly fix for https://github.com/paritytech/polkadot-sdk/issues/200

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2557572

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2558030

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2558413

File: primitives/arithmetic/src/per_things.rs [hunk 1]
```diff
@@ -506,6 +513,10 @@ where
 	let numer_upper = P::Upper::from(numer);
 	let denom_n: N = denom.into();
 	let denom_upper = P::Upper::from(denom);
+	if denom.is_zero() {
+		// Division by zero... best effort to prevent a panic.
+		return N::one()
+	}
 	let rem = x.rem(denom_n);
```
Comment (ggwpez) [inline primitives/arithmetic/src/per_things.rs:520]: The rem also panics for zero input.

**Comment by gavofyork** [comment]
Division by zero *should* panic if there is no way to return an `Err`. It panics in regular arithmetic. Returning some finite value could mask major errors (tests would no longer panic) and lead to unintended consequences.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2561621

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2561620

**Comment by ggwpez** [comment]
> Division by zero _should_ panic if there is no way to return an `Err`. It panics in regular arithmetic. Returning some finite, strictly incorrect, value could mask major errors (tests would no longer panic) and lead to unintended consequences.
> 
> It's reasonable to introduce alternative versions of these functions which return `Result` and thus need not panic for division-by-zero. If we really want to make the infallible functions not panic on division by zero, I would at least use `defensive!` to ensure the error does not go unnoticed.

Yea the case that it wont show up in tests anymore is indeed bad. So having `Result` functions is probably the only proper way since the code in `sp-*` cannot use the `defensive!` from `frame-support`. Should I create an issue for it?

File: primitives/arithmetic/src/per_things.rs (hunk 1 above)
Comment (ggwpez) [reply primitives/arithmetic/src/per_things.rs]: reverted since it wont show up in tests as panic

**Comment by gavofyork** [comment]
> > Division by zero _should_ panic if there is no way to return an `Err`. It panics in regular arithmetic. Returning some finite, strictly incorrect, value could mask major errors (tests would no longer panic) and lead to unintended consequences.
> > It's reasonable to introduce alternative versions of these functions which return `Result` and thus need not panic for division-by-zero. If we really want to make the infallible functions not panic on division by zero, I would at least use `defensive!` to ensure the error does not go unnoticed.
> 
> Yea the case that it wont show up in tests anymore is indeed bad. So having `Result` functions is probably the only proper way since the code in `sp-*` cannot use the `defensive!` from `frame-support`. Should I create an issue for it?

Yes, think so.

File: primitives/arithmetic/fuzzer/src/per_thing_mult_fraction.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	let q = P::from_rational_with_rounding(n, d, Rounding::Down).unwrap();
+	assert!(q * d <= n, "{:?} * {:?} <= {:?}", q, d, n);
+}
+
+/// Checks that `n / (n / d) >= d`.
+fn check_reciprocal_mul<P: PerThing>(n: u128, d: u128)
+where
+	P: PerThing + core::ops::Mul<u128, Output = u128>,
+{
+	let q = P::from_rational_with_rounding(n, d, Rounding::Down).unwrap();
```
Comment (ggwpez) [inline primitives/arithmetic/fuzzer/src/per_thing_mult_fraction.rs:62]: Do you know if the rounding here is correct?

**Review by ggwpez — APPROVED** [review]
% the comment on `multiply_rational`.  
Still need to figure out how to run these tests in the CI.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2562643

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2562642

File: primitives/arithmetic/src/rational.rs [hunk 3]
```diff
@@ -284,6 +284,50 @@ impl PartialEq for Rational128 {
...
+pub trait MultiplyRational: Sized {
+	fn multiply_rational(self, n: Self, d: Self, r: Rounding) -> Option<Self>;
+}
+
+macro_rules! impl_rrm {
+	($ulow:ty, $uhi:ty) => {
+		impl MultiplyRational for $ulow {
+			fn multiply_rational(self, n: Self, d: Self, r: Rounding) -> Option<Self> {
+				let sn = (self as $uhi) * (n as $uhi);
+				let mut result = sn / (d as $uhi);
```
Comment (ggwpez) [inline primitives/arithmetic/src/rational.rs:296]: Should return `None` now, or?

File: primitives/arithmetic/fuzzer/src/per_thing_mult_fraction.rs (hunk 2 above)
Comment (gavofyork) [reply primitives/arithmetic/fuzzer/src/per_thing_mult_fraction.rs:62]: doesn't look like it. you can't round down on both and make any expectations on overall rounding since they will work in opposite directions.

**Review by ruseinov — COMMENTED** [review]
the logic looks good to me, however we should probably fix the staking test to actually reflect this properly and not via magic numbers.

**Review by ruseinov — APPROVED** [review]
approving this as fixing the staking test properly could be done as part of another issue

**Comment by gavofyork** [comment]
bot merge
